apps/i18n.py

```diff
@@ -0,0 +1,178 @@
+                        sys.stderr.write("Error: Redefinition of message \"" + name + "\" as universal\n")
+                        sys.exit(-1)
+                    if name in universal_messages:
+                        sys.stderr.write("Error: Redefinition of universal message \"" + name + "\"\n")
+                        sys.exit(-1)
+                    universal_messages.add(name)
+                else:
+                    messages.add(name)
+                data[locale][name] = definition
+    return {"messages": sorted(messages), "universal_messages": sorted(universal_messages), "data": data}
+
+def print_header(data, path, locales):
+    f = open(path, 'w')
+    f.write("#ifndef APPS_I18N_H\n")
+    f.write("#define APPS_I18N_H\n")
+    f.write("\n")
+    f.write("// This file is auto-generated by i18n.py\n")
+    f.write("\n")
+    f.write("#include <escher.h>\n")
+    f.write("\n")
+    f.write("namespace I18n {\n")
+    f.write("\n")
+    f.write("constexpr static int NumberOfLanguages = %d;\n" % len(locales))
+    f.write("\n")
+    f.write("enum class Message : uint16_t {\n")
+    for message in data["messages"]:
+        f.write("  " + message + ",\n")
+    f.write("\n")
+    f.write("  UniversalMessageMarker,\n")
+    f.write("\n")
+    for message in data["universal_messages"]:
+        f.write("  " + message + ",\n")
+    f.write("};\n")
+    f.write("\n")
+    f.write("enum class Language : uint16_t {\n")
+    f.write("  Default = 0,\n")
+    for locale in locales:
+        f.write("  " + locale.upper() + ",\n")
+    f.write("};\n")
+    f.write("\n")
+    f.write("constexpr const Message LanguageNames[NumberOfLanguages] = {\n");
+    for locale in locales:
+        f.write("  Message::Language" + locale.upper() + ",\n")
+    f.write("};\n")
+    f.write("\n")
+    f.write("}\n")
+    f.write("\n")
+    f.write("#endif\n")
+    f.close()
+
+def print_implementation(data, path, locales):
+    f = open(path, 'w')
+    f.write("#include \"i18n.h\"\n")
+    f.write("#include \"global_preferences.h\"\n")
+    f.write("#include <assert.h>\n");
+    f.write("\n")
+    f.write("namespace I18n {\n")
+    f.write("\n")
+    for message in data["messages"]:
+        for locale in locales:
+            if not locale in data["data"]:
+                sys.stderr.write("Error: Undefined locale \"" + locale + "\"\n")
+                sys.exit(-1)
+            if not message in data["data"][locale]:
+                sys.stderr.write("Error: Undefined key \"" + message + "\" for locale \"" + locale + "\"\n")
+                sys.exit(-1)
+            f.write("constexpr static char " + locale + message + "[] = " + data["data"][locale][message] + ";\n")
+    f.write("\n")
+    f.write("constexpr static const char * messages[%d][%d] = {\n" % (len(data["messages"]), len(locales)))
+    for message in data["messages"]:
+        f.write("  {")
+        for locale in locales:
+            if not locale in data["data"]:
```

Comment:
Duplicated and redundant tests (the first `for` loop already cover these cases).